client/services/code_preprocessors/PythonCodePreprocessorService.js

```diff
@@ -126,11 +129,32 @@ tie.factory('PythonCodePreprocessorService', [
         while (true) {
           var codeToAdd = '';
           lastFunctionNameLocation = code.indexOf(
-            functionName, lastFunctionNameLocation);
+            functionNameWithoutParen, lastFunctionNameLocation);
           if (lastFunctionNameLocation === -1) {
             break;
           }
           if (lastFunctionNameLocation !== functionNameStart) {
+            var whitespaceCheckLocation = (
+              lastFunctionNameLocation + functionNameWithoutParen.length);
+            // We assume we found the matched function, but we might have found
+            // a similarly-named other function / variable. User-inputted code
+            // can be ... tricky.
+            var isMatchedFunction = true;
+            // We can safely assume there's a left paren because we matched it
+            // with the regular expression.
+            while (code[whitespaceCheckLocation] != '(') {
+              var isValidWhitespace = VALID_WHITESPACE_REGEX.test(
```

Comment:
I was confused about this for a while as well (and honestly still am, so I had to re-test the code), but here's my best explanation and let me know if this doesn't make sense.

The matching string is "match anything that looks like a python function definition _and_ ends with a parenthesis". The problem is that we might capture:

"def encode     ("

And would strip that down to "encode". That's perfectly fine. But, when we start searching for the string in the code we can't do an exact match on the function name, because we might have "encoded()" as a method, and then we would append the StudentCode(). wrapper twice. On the other hand, we can't do an exact match on the function name with the parenthesis because we might have "encode(", "encode (", and "encode   (" in the code, which I guess are all valid Python?

It's only a problem if a function name is a complete substring of another function name, which is uncommon but should still be supported, I guess.